sql: TestRaceWithBackfill failed [delete fails because batch timestamp must be after replica GC threshold]

sql.TestRaceWithBackfill [failed](https://tanzanite.cluster.engflow.com/invocations/default/092f05b7-b7cf-41a8-8de5-3bb3005e52e8?testReportRun=2&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on release-24.3 @ [4cbedefd790c75cb0f21f77ed8d917c8528a7d15](https://github.com/cockroachdb/cockroach/commits/4cbedefd790c75cb0f21f77ed8d917c8528a7d15):


```
=== RUN   TestRaceWithBackfill
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestRaceWithBackfill1823756680
    test_log_scope.go:76: use -show-logs to present logs inline
    schema_changer_test.go:341: schema change ALTER TABLE t.test ADD COLUMN x DECIMAL DEFAULT (DECIMAL '1.4') CHECK (x >= 0) took 16.264911616s
    schema_changer_test.go:341: schema change ALTER TABLE t.test DROP pi took 59.099274135s
    schema_changer_test.go:341: schema change CREATE UNIQUE INDEX foo ON t.test (v) took 53.127756921s
    schema_changer_test.go:341: schema change CREATE INDEX bar ON t.test(k) STORING (v) took 35.585343181s
    schema_changer_test.go:427: pq: batch timestamp 1729337275.789296372,0 must be after replica GC threshold 1729337288.197835522,0 (r73: /Table/106/{1/160-4/40})
    schema_changer_test.go:427: pq: batch timestamp 1729337304.143614837,0 must be after replica GC threshold 1729337309.768331147,0 (r111: /Table/106/1{0/160-1/40})
    schema_changer_test.go:734: e = 201, v = 208
    schema_changer_test.go:734: e = 202, v = 209
    schema_changer_test.go:745: read the wrong number of rows: e = 201, v = 203
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestRaceWithBackfill1823756680
--- FAIL: TestRaceWithBackfill (325.68s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRaceWithBackfill.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-43384

--- a/pkg/sql/schema_changer_test.go
+++ b/pkg/sql/schema_changer_test.go
@@ -563,6 +563,7 @@ func TestRaceWithBackfill(t *testing.T) {
 	defer log.Scope(t).Close(t)
 
 	skip.UnderDeadlock(t, "very long-running test under deadlock")
+	skip.UnderRace(t, "can cause flakes due to queries failing because of aggressive GC TTL")
 
 	// protects backfillNotification
 	var mu syncutil.Mutex
@@ -630,10 +631,11 @@ CREATE UNIQUE INDEX vidx ON t.test (v);
 		t.Fatal(err)
 	}
 	tableDesc := desctestutils.TestingGetPublicTableDescriptor(kvDB, keys.SystemSQLCodec, "t", "test")
-	// Disable strict GC TTL enforcement so that we can use AddImmediateGCZoneConfig.
+	// We are reducing the GC TTL to a low value and, as a precaution, disabling
+	// strict GC TTL enforcement. Previously, we made it immediate but occasionally
+	// encountered errors where the batch timestamp was before the replica GC threshold.
 	defer sqltestutils.DisableGCTTLStrictEnforcement(t, sqlDB)()
-	// Add a zone config for the table so that garbage collection happens rapidly.
-	if _, err := sqltestutils.AddImmediateGCZoneConfig(sqlDB, tableDesc.GetID()); err != nil {
+	if _, err := sqltestutils.UpdateGCZoneConfig(sqlDB, tableDesc.GetID(), 1); err != nil {
 		t.Fatal(err)
 	}
 	// Bulk insert.

--- a/pkg/sql/sqltestutils/sql_test_utils.go
+++ b/pkg/sql/sqltestutils/sql_test_utils.go
@@ -33,8 +33,17 @@ import (
 // AddImmediateGCZoneConfig set the GC TTL to 0 for the given table ID. One must
 // make sure to disable strict GC TTL enforcement when using this.
 func AddImmediateGCZoneConfig(sqlDB *gosql.DB, id descpb.ID) (zonepb.ZoneConfig, error) {
+	return UpdateGCZoneConfig(sqlDB, id, 0)
+}
+
+// UpdateGCZoneConfig sets the GC TTL to a custom value for the given table ID.
+// If setting the value to 0, one must make sure to disable strict GC TTL
+// enforcement when using this.
+func UpdateGCZoneConfig(
+	sqlDB *gosql.DB, id descpb.ID, ttlSeconds int32,
+) (zonepb.ZoneConfig, error) {
 	cfg := zonepb.DefaultZoneConfig()
-	cfg.GC.TTLSeconds = 0
+	cfg.GC.TTLSeconds = ttlSeconds
 	buf, err := protoutil.Marshal(&cfg)
 	if err != nil {
 		return cfg, err

